container/docker/factory.go

```diff
@@ -197,6 +208,80 @@ func startThinPoolWatcher(dockerInfo *dockertypes.Info) (*devicemapper.ThinPoolW
 	return thinPoolWatcher, nil
 }
 
+func getKernelVersion() (string, error) {
+	unameOutput := syscall.Utsname{}
+	err := syscall.Uname(&unameOutput)
+	if err != nil {
+		return "", err
+	}
+
+	b := make([]byte, len(unameOutput.Release))
+	for i, v := range unameOutput.Release {
+		b[i] = byte(v)
+	}
+	return string(b), nil
+}
+
+func ensureThinLsKernelVersion(kernelVersion string) error {
+	// kernel 4.4.0 has the proper bug fixes to allow thin_ls to work without corrupting the thin pool
+	version_4_4_0 := semver.MustParse("4.4.0")
+	// RHEL kernel 3.10.0 release >= 366 has the proper bug fixes backported from 4.4.0 to allow
+	// thin_ls to work without corrupting the thin pool
+	version_3_10_0 := semver.MustParse("3.10.0")
+
+	matches := version_re.FindStringSubmatch(kernelVersion)
+	if len(matches) < 4 {
+		return fmt.Errorf("error parsing kernel version %q: expected 4 matches, got %d", kernelVersion, len(matches))
+	}
+
+	sem, err := semver.Make(matches[0])
+	if err != nil {
+		return err
+	}
+
+	if sem.GTE(version_4_4_0) {
+		// kernel 4.4+ - good
+		return nil
+	}
+
+	// Certain RHEL/Centos 7.x kernels have a backport to fix the corruption bug
+	if !strings.Contains(kernelVersion, ".el7.") {
+		// not a RHEL 7.x kernel - won't work
+		return fmt.Errorf("kernel version 4.4.0 or later is required to use thin_ls - you have %q", kernelVersion)
+	}
+
+	// RHEL/Centos 7.x from here on
+	if sem.Major != 3 {
+		// only 3.x kernels *may* work correctly
```

Comment:
I can't speak to if any other vendors backported the bug fixes to something earlier than 4.4.0. (@eparis where did you find that 4.4.0 was the min version with the fix?) My logic flow was this:

1. Regardless of distro, if >= 4.4.0, good
1. Otherwise < 4.4.0
  1. If kernel doesn't contain ".el7.", bad
  1. Otherwise it does have ".el7", so:
    1. RHEL only backported it to 3.x, so if Major != 3, bad
    1. Major is 3, so check 3.10.0-366 or newer -> good